### PR TITLE
Add metric output support

### PR DESCRIPTION
**Pull request by pakrym**
@NTaylorMullen

### CODE REVIEW

File: src/Stress.Framework/StressRunSummary.cs [hunk 1]
```diff
@@ -38,10 +39,16 @@ public void PopulateMetrics(IStressMetricCollector collector)
             RequestsPerSecond = RequestCount / elapsedSeconds;
         }
 
-        public void PublishOutput()
+        public void PublishOutput(ITestCase testCase, IMessageBus messageBus)
         {
+            messageBus.QueueMessage(new StressTestStatisticsMessage(testCase, "ITER", Iterations));
```
Comment (pakrym) [inline src/Stress.Framework/StressRunSummary.cs]: Not sure if this belongs here

File: src/Stress.Framework/StressTestCaseRunner.cs [hunk 2]
```diff
@@ -231,6 +231,26 @@ protected override object CreateTestClass()
 
                 return testClass;
             }
+
+            protected override Task BeforeTestMethodInvokedAsync()
+            {
+                var stressTestCase = this.TestCase as StressTestCase;
```
Comment (NTaylorMullen) [inline src/Stress.Framework/StressTestCaseRunner.cs]: TestCase is overridden. No need to cast here: https://github.com/aspnet/Performance/blob/cae310e176e91f4e61e73f1cc279a1520ebaea63/src/Stress.Framework/StressTestCaseRunner.cs#L48

File: src/Stress.Framework/StressTestCaseRunner.cs [hunk 3]
```diff
@@ -231,6 +231,26 @@ protected override object CreateTestClass()
...
+                if (stressTestCase != null)
+                {
+                    stressTestCase.MetricReporter?.Start(MessageBus, stressTestCase);
```
Comment (NTaylorMullen) [inline src/Stress.Framework/StressTestCaseRunner.cs]: Lets assume `MetricReporter` is always set. As it stands right now it always will be.

File: src/Stress.Framework/StressTestCaseRunner.cs [hunk 4]
```diff
@@ -231,6 +231,26 @@ protected override object CreateTestClass()
...
+                }
+                return base.BeforeTestMethodInvokedAsync();
+            }
+
+            protected override Task AfterTestMethodInvokedAsync()
+            {
+                var testCase = this.TestCase as StressTestCase;
```
Comment (NTaylorMullen) [inline src/Stress.Framework/StressTestCaseRunner.cs]: ditto comments as in the BeforeTestMethod variation.

File: src/Stress.Framework/StressTeamCityReporter.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (NTaylorMullen) [inline src/Stress.Framework/StressTeamCityReporter.cs]: header

File: src/Stress.Framework/StressTestStatisticsMessage.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Stress.Framework
+{
+    class StressTestStatisticsMessage : TestCaseMessage, ITestStatisticsMessage
```
Comment (NTaylorMullen) [inline src/Stress.Framework/StressTestStatisticsMessage.cs]: public class

File: src/Stress.Framework/StressTestTeamCityReporterMessageHandler.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        public StressTestTeamCityReporterMessageHandler(IRunnerLogger logger,
+            Func<string, string> flowIdMapper = null,
+            TeamCityDisplayNameFormatter displayNameFormatter = null) : base(logger, flowIdMapper, displayNameFormatter)
+        {
+            _logger = logger;
+
+            // I know, I know, but I want them so bad and ITestProgressMessage and ITestStatisticsMessage won't
+            // be written too often, so should not hurt repformance.
+
+            var escapeMethodInfo = typeof(TeamCityReporterMessageHandler).GetMethod("Escape", BindingFlags.Static | BindingFlags.NonPublic);
```
Comment (NTaylorMullen) [inline src/Stress.Framework/StressTestTeamCityReporterMessageHandler.cs]: lol this doesn't work on CoreCLR does it? Wasn't aware BindingFlags was carried over.

File: src/Stress.Framework/StressTestCaseRunner.cs (hunk 2 above)
Comment (pakrym) [inline src/Stress.Framework/StressTestCaseRunner.cs]: This is StressTestInvoker, its not overriden here

File: src/Stress.Framework/StressTestTeamCityReporterMessageHandler.cs (hunk 7 above)
Comment (pakrym) [inline src/Stress.Framework/StressTestTeamCityReporterMessageHandler.cs]: Woks ok on coreclr

File: stress-test/Microsoft.AspNet.Tests.Stress/config.json [hunk 8]
```diff
@@ -1,6 +1,7 @@
 ﻿{
   "Stress": {
     "RunIterations": false,
-    "Iterations": 1000000
+    "Iterations": 1000000,
+    "MetricReportInterval": 1
```
Comment (NTaylorMullen) [inline stress-test/Microsoft.AspNet.Tests.Stress/config.json]: Make default 5

File: src/Stress.Framework/StressTestCaseRunner.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Stress.Framework/StressTestCaseRunner.cs]: Haha, damn GitHub diff :+1:

File: src/Stress.Framework/StressRunSummary.cs [hunk 9]
```diff
@@ -38,8 +39,14 @@ public void PopulateMetrics(IStressMetricCollector collector)
...
+            messageBus.QueueMessage(new StressTestStatisticsMessage(testCase, "TIME", TimeElapsed.Seconds));
+            messageBus.QueueMessage(new StressTestStatisticsMessage(testCase, "MEM", MemoryDelta / 1000));
+            messageBus.QueueMessage(new StressTestStatisticsMessage(testCase, "REQ", RequestCount));
+            messageBus.QueueMessage(new StressTestStatisticsMessage(testCase, "RPS", RequestsPerSecond));
+
             Console.WriteLine("Iterations: " + Iterations);
```
Comment (NTaylorMullen) [inline src/Stress.Framework/StressRunSummary.cs]: Would be nice if we could just replace the reporter to be a console writeline reporter as a replacement of these :smile:

**Comment by NTaylorMullen** [comment]
:shipit:

**Pull request merged**